server/controllers/UserController.js

```diff
@@ -32,22 +45,93 @@ class UserController {
           new: true, // Return the updated user
         },
       )
-      return updatedUser
+      return updatedUser.toObject()
     } catch (error) {
       console.error('Error while updating user:', error)
       throw error
     }
   }
 
-  async deleteUser(userId) {
+  async delete(userId) {
     try {
       const deletedUser = await UserModel.findByIdAndRemove(userId)
-      return deletedUser
+      return deletedUser.toObject()
     } catch (error) {
       console.error('Error while deleting user:', error)
       throw error
     }
   }
+
+  async register(userData) {
+    const userWithEmail = await this.getByEmail(userData.email)
+    if (userWithEmail != null) {
+      const error = new Error('Email already exists')
+      error.statusCode = 400
+      throw error
+    }
+    const salt = bcrypt.genSaltSync(10)
+    const hashedPassword = bcrypt.hashSync(userData.password, salt)
+    const newlyCreatedUser = await this.create({
+      username: userData.username,
+      email: userData.email,
+      password: hashedPassword,
+    })
+
+    const accessToken = this.createAccessToken(newlyCreatedUser)
+    const userDTO = {
+      id: newlyCreatedUser._id,
+      username: newlyCreatedUser.username,
+      email: newlyCreatedUser.email,
+      accessToken,
+    }
+    return userDTO
+  }
+
+  createAccessToken(userData) {
+    try {
+      const accessToken = jwt.sign(
+        {
+          id: userData._id,
+          username: userData.username,
+          email: userData.email,
+        },
+        config.server.jwtSecret,
+      )
+      return accessToken
+    } catch (error) {
+      console.error('Error with creating token')
+      throw new Error('Could not create token: ', error)
+    }
+  }
+
+  async login(userData) {
+    const userWithEmail = await this.getByEmail(userData.email)
+    if (userWithEmail == null) {
+      const error = new Error('No user with this email')
+      error.statusCode = 404
+      throw error
+    }
+
+    const passwordMatches = bcrypt.compareSync(
+      userData.password,
+      userWithEmail.password,
+    )
+
+    if (!passwordMatches) {
+      const error = new Error('Invalid password')
+      error.statusCode = 400
+      throw error
+    }
+
+    const accessToken = this.createAccessToken(userWithEmail)
+    const userDTO = {
+      id: userWithEmail._id,
+      username: userWithEmail.username,
+      email: userWithEmail.email,
+      accessToken,
+    }
+    return userDTO
```

Comment:
I may be misunderstanding something but why do we return the DTO instead of just the access token? What value does sending the id/username/email give?

Reply:
returning more than just the access token provides the user with additional information about the authenticated user right after registration or login without additional request to fetch

Reply:
ah I see. where might they use this additional information? username I can see used in the "logout X" button but I don't know for id and email

Reply:
in our app, i dont think we need it. so probably can remove it if u want and add it back if we need it.

Reply:
lol ty